### PR TITLE
Avoid redundant callback restore warning while tuning

### LINKED ISSUE #12640: Warning of ambiguous batch size when using torchmetrics
## 🐛 Bug

When I use torchmetrics with some "ambiguous batch size" data, the batch size for each update step can be handled by using torchmetrics API, so pytorch lightning does not need to know the batch size. But lightning always gives me warnings like "UserWarning: Trying to infer the `batch_size` from an ambiguous collection. The batch size we found is 1. To avoid any miscalculations, use `self.log(..., batch_size=batch_size)`."

Sorry I don't much time to provide codes for this. I think this might be a bug and hope that you can fix it.

cc @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy

### LINKED ISSUE #12817: "Please add the following callbacks" warning
## 🐛 Bug

Pytorch Lightning emits a false positive (?) warning when restoring from checkpoint during model tuning with `auto_lr_find`.

### To Reproduce

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import ModelCheckpoint


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self, lr=0.1):
        self.save_hyperparameters()
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=self.hparams.lr)

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 1024), batch_size=2, num_workers=8)

    def val_dataloader(self):
        return DataLoader(RandomDataset(32, 1024), batch_size=2, num_workers=8)

    def test_dataloader(self):
        return DataLoader(RandomDataset(32, 1024), batch_size=2, num_workers=8)



def run():
    model = BoringModel()

    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=1,
        auto_lr_find=True,
        callbacks=[ModelCheckpoint()]
    )
    trainer.tune(model, lr_find_kwargs={"early_stop_threshold": None})


if __name__ == "__main__":
    run()
```

### Expected behavior

No warnings

### Actual behavior

```
/usr/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py:1721: UserWarning: Be aware that when using `ckpt_path`, callbacks used to create the checkpoint need to be provided during `Trainer` instantiation. Please add the following callbacks: ["ModelCheckpoint{'monitor': None, 'mode': 'min', 'every_n_train_steps': 0, 'every_n_epochs': 1, 'train_time_interval': None, 'save_on_train_epoch_end': None}"].
```

### Environment

```
* CUDA:
	- GPU:
		- NVIDIA GeForce GTX 1050
	- available:         True
	- version:           11.3
* Packages:
	- numpy:             1.22.3
	- pyTorch_debug:     False
	- pyTorch_version:   1.11.0+cu113
	- pytorch-lightning: 1.6.0
	- tqdm:              4.63.2
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         
	- python:            3.10.4
	- version:           #1 SMP PREEMPT Mon Mar 28 09:16:36 UTC 2022
```

cc @otaj @akihironitta @borda @rohitgr7

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #12817

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @otaj @akihironitta @borda @rohitgr7

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -139,7 +139,8 @@ def restore(self, checkpoint_path: Optional[_PATH] = None) -> None:
         self.restore_model()
 
         # restore callback states
-        self.restore_callbacks()
+        if self.trainer.state.fn != TrainerFn.TUNING:
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:142]: I'm confused actually. The other place `restore_callabacks` is called uses `fit`: https://github.com/PyTorchLightning/pytorch-lightning/blob/3f78c4ca7a89fed8f79d9015e9510f0a0dcc46b5/pytorch_lightning/trainer/trainer.py#L1101

But why would callbacks not be restored for all trainer functions?

File: tests/tuner/test_lr_finder.py [hunk 2]
```diff
@@ -87,9 +88,11 @@ def test_trainer_reset_correctly(tmpdir):
         "max_steps",
     ]
     expected = {ca: getattr(trainer, ca) for ca in changed_attributes}
-    trainer.tuner.lr_find(model, num_training=5)
-    actual = {ca: getattr(trainer, ca) for ca in changed_attributes}
 
+    with no_warning_call(UserWarning, match="Please add the following callbacks"):
```
Comment (carmocca) [inline tests/tuner/test_lr_finder.py:92]: The linked issue was closed saying that the warning disappeared. I would not make this change then

Comment (rohitgr7) [reply tests/tuner/test_lr_finder.py:92]: fixed the description. I linked another issue by mistake.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:142]: because we clear out all the callbacks while tuning and use checkpoint connector to restore states of everything (including the callbacks). The saved checkpoint is restored before reloading the params.

https://github.com/PyTorchLightning/pytorch-lightning/blob/3f78c4ca7a89fed8f79d9015e9510f0a0dcc46b5/pytorch_lightning/tuner/lr_finder.py#L207-L237

An alternate solution can be to restore the params and load the checkpoint.

**Review by carmocca — APPROVED** [review]
LGTM!

**Review by akihironitta — APPROVED** [review]
LGTM! 🚀

**Review by Borda — APPROVED** [review]
